src/main/java/io/atomix/client/impl/AbstractAsyncPrimitive.java

```diff
@@ -0,0 +1,189 @@
+// SPDX-FileCopyrightText: 2022-present Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package io.atomix.client.impl;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import io.atomix.api.runtime.v1.PrimitiveId;
+import io.atomix.client.AsyncPrimitive;
+import io.atomix.client.Cancellable;
+import io.atomix.client.iterator.AsyncIterator;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Simple asynchronous primitive.
+ */
+public abstract class AbstractAsyncPrimitive<P extends AsyncPrimitive> implements AsyncPrimitive {
```

Comment:
Another doubt I have from my past experience is about cancellable context. Do we need to handle that ? I dont see you handling it in this patch.